ir/src/lib.rs

```diff
@@ -71,38 +84,73 @@ impl AirIR {
 
         Ok(Self {
             air_name: air_name.to_string(),
-            main_boundary_constraints,
-            main_transition_constraints,
+            periodic_columns: symbol_table.into_periodic_columns(),
+            boundary_constraints,
+            transition_constraints,
         })
     }
 
     // --- PUBLIC ACCESSORS -----------------------------------------------------------------------
+
     pub fn air_name(&self) -> &str {
         &self.air_name
     }
 
+    pub fn periodic_cycle_lens(&self) -> Vec<usize> {
+        self.periodic_columns
+            .iter()
+            .map(|values| values.len())
+            .collect()
+    }
+
+    // --- PUBLIC ACCESSORS FOR BOUNDARY CONSTRAINTS ----------------------------------------------
+
     pub fn num_main_assertions(&self) -> usize {
-        self.main_boundary_constraints.len()
+        self.boundary_constraints.main_len()
+    }
+
+    pub fn main_first_boundary_constraints(&self) -> Vec<&BoundaryExpr> {
+        self.boundary_constraints.main_first()
     }
 
-    pub fn main_first_boundary_constraints(&self) -> Vec<&Expr> {
-        self.main_boundary_constraints.first()
+    pub fn main_last_boundary_constraints(&self) -> Vec<&BoundaryExpr> {
+        self.boundary_constraints.main_last()
     }
 
-    pub fn main_last_boundary_constraints(&self) -> Vec<&Expr> {
-        self.main_boundary_constraints.last()
+    pub fn num_aux_assertions(&self) -> usize {
+        self.boundary_constraints.aux_len()
     }
 
-    pub fn main_degrees(&self) -> Vec<u8> {
-        self.main_transition_constraints.degrees()
+    pub fn aux_first_boundary_constraints(&self) -> Vec<&BoundaryExpr> {
+        self.boundary_constraints.aux_first()
+    }
+
+    pub fn aux_last_boundary_constraints(&self) -> Vec<&BoundaryExpr> {
+        self.boundary_constraints.aux_last()
+    }
+
+    // --- PUBLIC ACCESSORS FOR TRANSITION CONSTRAINTS --------------------------------------------
+
+    pub fn main_degrees(&self) -> Vec<TransitionConstraintDegree> {
+        self.transition_constraints
+            .main_degrees(&self.periodic_cycle_lens())
     }
 
     pub fn main_transition_constraints(&self) -> &[NodeIndex] {
-        self.main_transition_constraints.constraints()
+        self.transition_constraints.main_constraints()
+    }
+
+    pub fn aux_degrees(&self) -> Vec<TransitionConstraintDegree> {
+        self.transition_constraints
+            .aux_degrees(&self.periodic_cycle_lens())
+    }
+
+    pub fn aux_transition_constraints(&self) -> &[NodeIndex] {
+        self.transition_constraints.aux_constraints()
     }
 
-    pub fn main_transition_graph(&self) -> &AlgebraicGraph {
-        self.main_transition_constraints.graph()
+    pub fn transition_graph(&self) -> &AlgebraicGraph {
```

Comment:
minor nit (feel free to ignore): I don't love the name `transition_graph`. Something like `transition_constraint_expressions` would be more correct, but it is rather long - so, don't love that either.

Reply:
I could call it `constraint_graph` if you prefer that?

Reply:
Not sure if it is better, to be honest :) I'm fine keeping it as is.